deepbots-panda to deepworlds

KelvinYang0320 opened the pull request.

KelvinYang0320 committed:
upload panda

update readme

update readme

update readme

del break

Update README.md

Update README.md

refactoring

refactoring

refactoring

forget to save my models!

os

os

tmp/ddpg

save models every 200 episodes

del boundingObject

del boundingObject

fix solved() SMA100->SMA500

fix maintainer comments

del #... in ddpg.py

add const

const:capital letters

update

tsampazk commented:
Hey @KelvinYang0320! First of all, thank you for this great addition to the repository. I would also like to commend you on the quality of the code, great job!

One issue i noticed and that seems to cause a significant slow down in the training process is that when the arm self-collides, the simulation slows down to a crawl. I am including some screenshots below.

![selfcollision](https://user-images.githubusercontent.com/27914645/110784018-d1efc880-8271-11eb-90b2-30cd8c402201.png)
![selfcollision2](https://user-images.githubusercontent.com/27914645/110784029-d4eab900-8271-11eb-95b4-e5d9649c17fb.png)
![nocollision](https://user-images.githubusercontent.com/27914645/110784040-d7e5a980-8271-11eb-9377-5f6aedab9331.png)
![warningPhysicsStep](https://user-images.githubusercontent.com/27914645/110784049-dae09a00-8271-11eb-957c-ddd8c51c66b8.png)


I guess that the the robot mesh becomes red when self-colliding and the simulation struggles to move forward, sometimes even showing the physics step warning. I am running the simulation in as-fast-as-possible speed and goes up to x15 when no self-collision is happening and slows down to x1.0 or less when self-colliding. Can you confirm that this is indeed happening on your system as well?

I can imagine a possible solution of detecting self-collision and add it as a termination (done) condition with some possible negative reward. It might greatly improve the speed of the training process. I can see that the [webots docs](https://cyberbotics.com/doc/reference/robot#field-summary) it states that the simulation might slow down a lot, but i am not exactly sure if there's a way to detect it in code. I posted a question in the Webots discord and will notify you here when i get an answer.

KelvinYang0320 commented:
Hello @tsampazk
Yes, this warning is sometimes happening on my system as well, but it only slows down the training process, so I ignore it.
The complex boundingObject is for grabing a beaker (my original training task).
In order to train faster, maybe we can just remove the boundingObject?
Without detecting self-collision, I think this task should still be able to be solved.

The gif below is the same goal reaching task that was solved in my [IKPY-based RL Env](https://github.com/KelvinYang0320/IKPY-based-RL-ENV), and this environment is without the self-collision and boundingObject.
![](https://i.imgur.com/St2TVSh.gif )

tsampazk commented:
I opened a feature request on the Webots repository, because it seems a great addition for RL tasks. For this PR maybe you can try removing the bounding box as you suggest as it is not critical. We can re-address this in the future, in relation to the feedback we will get from the Webots developers.

stefaniapedrazzi commented:
Just also note that one main problem of this robot model causing low simulation performances are the boundingObjects.
Using these complex IndexedFaceSet geometries is computationally expensive. It is strongly recommended to simplify them and when possible use only simple geometries (Spheres, Cylinder, Boxes, Capsules) in boundingObject definition.

tsampazk commented:
@all-contributors please add @KelvinYang0320 for code, doc, example, ideas

KelvinYang0320 commented:
> For this PR maybe you can try removing the bounding box as you suggest as it is not critical. 

@tsampazk 
Some parts of the robot arm will pass through itself during training by this solution.:sweat_smile:
Is that ok?

tsampazk commented:
> > For this PR maybe you can try removing the bounding box as you suggest as it is not critical.
> 
> @tsampazk
> Some parts of the robot arm will pass through itself during training by this solution.😅
> Is that ok?

Yes, i guess it's ok for the purposes of this example! 😄 You can go forward and remove it and we will add more thorough reviews and it should be good to go.

KelvinYang0320 commented:
@tsampazk Okay, I have removed it.
Thank you for adding me as a contributor!

KelvinYang0320 commented:
@tsampazk After training for about 12 hours, I think I should set the <code>env.solved()</code> condition to a higher score due to this [image](https://github.com/KelvinYang0320/deepworlds/blob/dev/examples/panda/doc/trend.png).
Are there other bugs I need to fix?

tsampazk commented:
> @tsampazk After training for about 12 hours, I think I should set the `env.solved()` condition to a higher score due to this [image](https://github.com/KelvinYang0320/deepworlds/blob/dev/examples/panda/doc/trend.png).

Since you have observed the results many more times and have more experience on the problem than me, if you think it is better this way, it looks good to me! 

As a minor note, i think that we would prefer to keep the solved conditions and hyperparameters in a state where it is as easy as possible for someone to train the agent out-of-the-box. For example in the find target problem we provide, the training procedure is problematic and we are looking into fixing it in the future.

> Are there other bugs I need to fix?

To be honest i tried running the example on my system and the robot controller crashes silently after 50-100 episodes. Obviously this is not happening to you, so i am wondering what Webots version are you using? Moreover, i think that we are both running the latest dev version of deepbots installed from [here](https://test.pypi.org/project/deepbots/), correct?

tsampazk commented:
@KelvinYang0320 Also you can take a look at the suggestion posted [here](https://github.com/cyberbotics/webots/issues/2846#issuecomment-796961095). Maybe it is something you can look into in the future if you find yourself working on a similar problem?

KelvinYang0320 commented:
> To be honest i tried running the example on my system and the robot controller crashes silently after 50-100 episodes. Obviously this is not happening to you, so i am wondering what Webots version are you using? Moreover, i think that we are both running the latest dev version of deepbots installed from [here](https://test.pypi.org/project/deepbots/), correct?

Yes, we are both running the latest dev version of deepbots with Webots R2021a, and I ran this example on Ubuntu 18.04 & Ubuntu 20.04. I have listed this information on the [READMD.md](https://github.com/KelvinYang0320/deepworlds/tree/dev/examples/panda#installation).
I think this is caused by memory limit (RAM, GPU)?

KelvinYang0320 commented:
> As a minor note, i think that we would prefer to keep the solved conditions and hyperparameters in a state where it is as easy as possible for someone to train the agent out-of-the-box. For example in the find target problem we provide, the training procedure is problematic and we are looking into fixing it in the future.

I can remove the 6 goals  from the 9 goals to simplify this problem.
> the robot controller crashes silently after 50-100 episodes.

However, since the robot controls 7 motor positions (in the huge action space), it still might not be able to be solved within 50-100 episodes.

> @KelvinYang0320 Also you can take a look at the suggestion posted [here](https://github.com/cyberbotics/webots/issues/2846#issuecomment-796961095). Maybe it is something you can look into in the future if you find yourself working on a similar problem?

Thank you very much for opening this issue on webots for me. :smile:
We have done a project with IKPY and DDPG, but we didn't use IKPY to simplify the action that the agent sends.
We separated a sequence of actions into many parts and solved each part by IKPY or DDPG.
I will look into IKFast Kinematics Solver in the future.:thinking:

tsampazk commented:
>  I have listed this information on the READMD.md.

Oops sorry i didn't notice that part of the README! I am having the issue on Windows 10.

> I think this is caused by memory limit (RAM, GPU)?

Ram usage seems stable, GPU is barely used, and the rest of the system runs fine. Even disabling the SYNCHRONIZATION flag of the controller enables the rest of the simulation to continue running. One thing i noticed is that it happened during the handshaking process. I tried to reproduce it a another time but it has reached 250 episodes and haven't happened. Really weird indeed.

@ManosMagnus could you please do a test run and see if the issue appears on your system as well when you get a chance? If my colleague Manos can't reproduce i think we can blame it on my system.

>However, since the robot controls 7 motor positions (in the huge action space), it still might not be able to be solved within 50-100 episodes.

Of course! Then we would have a pretty spectacular RL agent in our hands! 😛 You can keep it as it is, no problem at all. The fact that the provided code solves the problem is good enough. If you can find a way to make it faster it would be great, but it is not a must, so don't worry about it.

>Thank you very much for opening this issue on webots for me. 😄

You are welcome! Just food for thought for the future! 😁

KelvinYang0320 commented:
> I tried to reproduce it a another time but it has reached 250 episodes and haven't happened. Really weird indeed.

I also tried to reproduce the issue you mentioned on my Windows 10, but it has reached over 500 episodes and haven't happened.:thinking:

> You can keep it as it is, no problem at all.

Thank you! I think I will just keep it as it is.

eakirtas commented:
Hello @KelvinYang0320, 

Sorry for the delay, I will provide a code review on the following days

eakirtas commented on examples/panda/panda_goal_reaching/controllers/robot_supervisor_manager/DDPG_runner.py:
```diff
@@ -0,0 +1,85 @@
+              
+    # agent.load_models() # Load the pretrained model
+    episodeCount = 0 
+    episodeLimit = 50000
```
This might be a universal constant value or argument of run function?

KelvinYang0320 replied:
<code>episodeCount = 0</code> and <code>episodeLimit = 50000</code> are set as universal constants.
But we still can change these constants when doing other tasks.
Similar to these [lines](https://github.com/aidudezzz/deepworlds/blob/5f86cdd69b4e9906b43ccc4acc3d31b3e60125b9/examples/cartpole/cartpole_discrete/controllers/robot_supervisor_manager/PPO_runner.py#L16) in cartpole_discrete example.

eakirtas replied:
I meant that this can be on top of the file (bellow imports something like `EPISODE_LIMIT`) in order to be visible to the end user. Indeed this is as in cartpole, maybe we should change there as well. Of course, it is really minor, if you feel that's better as it is, feel free to resolve the comment

eakirtas commented on examples/panda/panda_goal_reaching/controllers/robot_supervisor_manager/DDPG_runner.py:
```diff
@@ -0,0 +1,85 @@
+    # We pass (10, ) as numberOfInputs and (7, ) as numberOfOutputs, taken from the gym spaces
+    agent = DDPGAgent(alpha=0.000025, beta=0.00025, input_dims=[env.observation_space.shape[0]], tau=0.001, batch_size=64,  layer1_size=400, layer2_size=400, n_actions=env.action_space.shape[0]) 
+              
+    # agent.load_models() # Load the pretrained model
```
Might be better if we have the option to pass a path at run function in order to load the pre-trained

KelvinYang0320 replied:
Ok, I will set an option to pass a path at run function.

eakirtas commented on examples/panda/panda_goal_reaching/controllers/robot_supervisor_manager/DDPG_runner.py:
```diff
@@ -0,0 +1,85 @@
+                # Save the episode's score
+                env.episodeScoreList.append(env.episodeScore)
+                agent.learn()
+                if episodeCount%200==0:
```
We can store `200` at a constant variable in order to be easily changed

KelvinYang0320 replied:
Ok

eakirtas commented on examples/panda/panda_goal_reaching/controllers/robot_supervisor_manager/DDPG_runner.py:
```diff
@@ -0,0 +1,85 @@
+    env.target = env.getFromDef("TARGET%s"%(random.randint(1, 10, 1)[0])) # Select one of the targets
+    while True:
+        act = agent.choose_action_test(state)
+        state, reward, done, _ = env.step(act*0.032)
```
Just a question: why you multiply action by 0.032?

KelvinYang0320 replied:
Change all actions from [-1,1] to [-0.032, 0.032] since I think that taking "1"or"-1" for actions is too large.

eakirtas commented on examples/panda/panda_goal_reaching/controllers/robot_supervisor_manager/DDPG_runner.py:
```diff
@@ -0,0 +1,85 @@
+
+            # Step the supervisor to get the current selectedAction reward, the new state and whether we reached the
+            # the done condition
+            newState, reward, done, info = env.step(act*0.032)
```
Just a question: why you multiply action by 0.032?

KelvinYang0320 replied:
Change all actions from [-1,1] to [-0.032, 0.032] since I think that taking "1"or"-1" for actions is too large.

eakirtas commented on examples/panda/panda_goal_reaching/controllers/robot_supervisor_manager/agent/ddpg.py:
```diff
@@ -0,0 +1,335 @@
+        #f2 = 0.002
+        T.nn.init.uniform_(self.fc2.weight.data, -f2, f2)
+        T.nn.init.uniform_(self.fc2.bias.data, -f2, f2)
+        #self.fc2.weight.data.uniform_(-f2, f2)
+        #self.fc2.bias.data.uniform_(-f2, f2)
```
Remove those lines?

KelvinYang0320 replied:
Ok

eakirtas commented on examples/panda/panda_goal_reaching/controllers/robot_supervisor_manager/agent/ddpg.py:
```diff
@@ -0,0 +1,335 @@
+        self.q = nn.Linear(self.fc2_dims, 1)
+        T.nn.init.uniform_(self.q.weight.data, -f3, f3)
+        T.nn.init.uniform_(self.q.bias.data, -f3, f3)
+        #self.q.weight.data.uniform_(-f3, f3)
+        #self.q.bias.data.uniform_(-f3, f3)
```
Remove those lines?

KelvinYang0320 replied:
Ok

eakirtas commented on examples/panda/panda_goal_reaching/controllers/robot_supervisor_manager/agent/ddpg.py:
```diff
@@ -0,0 +1,335 @@
+        f1 = 1./np.sqrt(self.fc1.weight.data.size()[0])
+        T.nn.init.uniform_(self.fc1.weight.data, -f1, f1)
+        T.nn.init.uniform_(self.fc1.bias.data, -f1, f1)
+        #self.fc1.weight.data.uniform_(-f1, f1)
+        #self.fc1.bias.data.uniform_(-f1, f1)
```
Remove those lines?

KelvinYang0320 replied:
Ok

eakirtas commented on examples/panda/panda_goal_reaching/controllers/robot_supervisor_manager/checkConvergence.py:
```diff
@@ -0,0 +1,23 @@
+def moving_average(x, w):
+    return np.convolve(x, np.ones(w), 'valid') / w
+
+fp=open("Episode-score.txt", "r")
+lines = fp.read().splitlines()
+scores = list(map(float, lines))
+episode = list(range(1, 1+len(scores)))
+print(scores)
+print(episode)
+	
+plt.figure()
+plt.title("Episode scores over episode")
+plt.plot(episode, scores, label='Raw data')
+SMA = moving_average(scores, 500)
+plt.plot(SMA, label='SMA500')
+plt.xlabel("episode")
+plt.ylabel("episode score")
+
+plt.legend()
+plt.savefig('trend.png')
+print("Last SMA500:",np.mean(scores[-500:]))
```
I think it will be better if we make include those lines in a function receiving arguments about the text file and the destination of plot

KelvinYang0320 replied:
Ok, check this [link](https://github.com/KelvinYang0320/deepworlds/blob/2af27e3dcd35d8bd8b305a0b480e64eab703c65b/examples/panda/panda_goal_reaching/controllers/robot_supervisor_manager/checkConvergence.py#L6), thanks!

eakirtas commented on examples/panda/panda_goal_reaching/controllers/robot_supervisor_manager/robot_supervisor_ddpg.py:
```diff
@@ -0,0 +1,225 @@
+        self.setup_motors()
+
+        # Set up misc
+        self.stepsPerEpisode = 300  # How many steps to run each episode (changing this messes up the solved condition)
```
Argument or constant?

KelvinYang0320 replied:
<code>self.stepsPerEpisode = 300</code> is a constant.
If user are trying to reach farther target, maybe user can change this constant.

eakirtas replied:
Similar as the previous comment. Feel free resolve the comment if need be

eakirtas commented on examples/panda/panda_goal_reaching/controllers/robot_supervisor_manager/robot_supervisor_ddpg.py:
```diff
@@ -0,0 +1,225 @@
+        self.stepsPerEpisode = 300  # How many steps to run each episode (changing this messes up the solved condition)
+        self.episodeScore = 0  # Score accumulated during an episode
+        self.episodeScoreList = []  # A list to save all the episode scores, used to check if task is solved
+        self.motorVelocity = 10
```
Argument or constant?

KelvinYang0320 replied:
Similar to these [lines](https://github.com/aidudezzz/deepworlds/blob/5f86cdd69b4e9906b43ccc4acc3d31b3e60125b9/examples/cartpole/cartpole_discrete/controllers/robot_supervisor_manager/robot_supervisor.py#L77) in cartpole_discrete example.
I set these as constants, and users can modify these lines if they want.

eakirtas commented on examples/panda/panda_goal_reaching/controllers/robot_supervisor_manager/checkConvergence.py:
```diff
@@ -0,0 +1,23 @@
+def moving_average(x, w):
+    return np.convolve(x, np.ones(w), 'valid') / w
+
+fp=open("Episode-score.txt", "r")
```
Might something like `exports/Episode-score.txt`

KelvinYang0320 replied:
Ok, I will change dest for <code>exports/Episode-score.txt</code>.

eakirtas reviewed (changes requested):
Look good to me. I've run example without any problem. Some minor comments were added.

KelvinYang0320 commented:
@ManosMagnus Thank you for your comments!
I have fixed these parts you mention.
If there are still some problems, plz let me know.:grinning:

eakirtas reviewed (commented):
LGTM, just two minors comments. Could you please [squash](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) the commits in order to get merged?

KelvinYang0320 commented:
@ManosMagnus I have set <code>STEPS_PER_EPISODE</code> and <code>MOTRO_VELOCITY</code> in [_robot_supervisor_ddpg.py_](https://github.com/KelvinYang0320/deepworlds/blob/507f2f366d9c8438257b9ef5130c4d12b3dbadb7/examples/panda/panda_goal_reaching/controllers/robot_supervisor_manager/robot_supervisor_ddpg.py#L5) as constants and set <code>EPISODE_LIMIT</code> and <code>SAVE_MODELS_PERIOD</code> as constants in [_DDPG_runner.py_](https://github.com/KelvinYang0320/deepworlds/blob/507f2f366d9c8438257b9ef5130c4d12b3dbadb7/examples/panda/panda_goal_reaching/controllers/robot_supervisor_manager/DDPG_runner.py#L6).
Thank you for your clear explanation!:sweat_smile:

Do you think we should create a python file, say _constant.py_, for these constants?
I need to use [<code>from robot_supervisor_ddpg import STEPS_PER_EPISODE</code>](https://github.com/KelvinYang0320/deepworlds/blob/507f2f366d9c8438257b9ef5130c4d12b3dbadb7/examples/panda/panda_goal_reaching/controllers/robot_supervisor_manager/DDPG_runner.py#L6) in _DDPG_runner.py_ since I need this constant from different python file.

eakirtas commented:
Thank you @KelvinYang0320!

That's a good idea. However, I am afraid that it could be not visible to the end user. Do you think constants is better to be added at [`robot_supervisor_manager.py`](https://github.com/aidudezzz/deepworlds/pull/33/files#diff-265454c8063017700997530c6712b002e4f79d4ee56f11e768712ebb0c36c8c8)

KelvinYang0320 commented:
Sorry for the delay in responding.
>  Do you think constants is better to be added at [`robot_supervisor_manager.py`](https://github.com/aidudezzz/deepworlds/pull/33/files#diff-265454c8063017700997530c6712b002e4f79d4ee56f11e768712ebb0c36c8c8)

Yes, that is a good idea. I will modify it.

KelvinYang0320 commented:
@ManosMagnus Not sure why I get these error if I import these constants from <code>robot_supervisor_manager.py</code>
![errConst](https://user-images.githubusercontent.com/49781698/113740436-457edb80-9733-11eb-9d3d-6868e3001c95.PNG)
But no error if I import these constants from <code>Constants.py</code>:confused:
You can reproduce this error by change <code>Constants</code> with <code>robot_supervisor_manager</code> in [this line](https://github.com/KelvinYang0320/deepworlds/blob/45cb3e96f02bb59209a177ff74c0e26b9d3907f3/examples/panda/panda_goal_reaching/controllers/robot_supervisor_manager/robot_supervisor_ddpg.py#L6) or [this line](https://github.com/KelvinYang0320/deepworlds/blob/45cb3e96f02bb59209a177ff74c0e26b9d3907f3/examples/panda/panda_goal_reaching/controllers/robot_supervisor_manager/DDPG_runner.py#L6).

eakirtas commented:
Might be because of the cross imports.  We can just keep it on `constants.py`.

I think that can merge the PR for now. Please remove and constants from manager and squash the commit in order to merge the PR.

Thank you!

KelvinYang0320 commented:
>  We can just keep it on `constants.py`.

> Please remove and constants from manager and squash the commit in order to merge the PR.

Ok, I have removed them and squashed all commits to one commit.
Thank you.

eakirtas commented:
Thank you @KelvinYang0320 for contributing such a nice example!
